Python/ceval.c

```diff
@@ -1924,6 +1926,82 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
             DISPATCH();
         }
 
+        case TARGET(BINARY_SUBSCR_ADAPTIVE): {
+            SpecializedCacheEntry *cache = GET_CACHE();
+            if (cache->adaptive.counter == 0) {
+                PyObject *sub = TOP();
+                PyObject *container = SECOND();
+                next_instr--;
+                if (_Py_Specialize_BinarySubscr(container, sub, next_instr, cache) < 0) {
+                    goto error;
+                }
+                DISPATCH();
+            }
+            else {
+                STAT_INC(BINARY_SUBSCR, deferred);
+                cache->adaptive.counter--;
+                JUMP_TO_INSTRUCTION(BINARY_SUBSCR);
+            }
+        }
+
+        case TARGET(BINARY_SUBSCR_LIST): {
+            PyObject *sub = TOP();
+            PyObject *list = SECOND();
+            DEOPT_IF(!PyLong_CheckExact(sub), BINARY_SUBSCR);
+            DEOPT_IF(!PyList_CheckExact(list), BINARY_SUBSCR);
+            Py_ssize_t index = PyLong_AsSsize_t(sub);
+            DEOPT_IF(index < 0 || index >= PyList_Size(list), BINARY_SUBSCR);
+            STAT_INC(BINARY_SUBSCR, hit);
+            STACK_SHRINK(1);
+            PyObject *res = PyList_GET_ITEM(list, index);
+            Py_XINCREF(res);
+            Py_DECREF(list);
+            Py_DECREF(sub);
+            SET_TOP(res);
+            if (res == NULL)
+                goto error;
+            DISPATCH();
+        }
+
+        case TARGET(BINARY_SUBSCR_TUPLE): {
+            PyObject *sub = TOP();
+            PyObject *tuple = SECOND();
+            DEOPT_IF(!PyLong_CheckExact(sub), BINARY_SUBSCR);
+            DEOPT_IF(!PyTuple_CheckExact(tuple), BINARY_SUBSCR);
+            Py_ssize_t index = PyLong_AsSsize_t(sub);
+            DEOPT_IF(index < 0 || index >= PyTuple_Size(tuple), BINARY_SUBSCR);
+            STAT_INC(BINARY_SUBSCR, hit);
+            STACK_SHRINK(1);
+            PyObject *res = PyTuple_GET_ITEM(tuple, index);
+            Py_XINCREF(res);
+            Py_DECREF(tuple);
+            Py_DECREF(sub);
+            SET_TOP(res);
+            if (res == NULL)
+                goto error;
+            DISPATCH();
+        }
+
+        case TARGET(BINARY_SUBSCR_DICT): {
```

Comment:
You can reduce the number of branches here by testing `res` before the incrementing its refcount, and by moving the error handling out of line.
```C
case TARGET(BINARY_SUBSCR_DICT): {
    PyObject *dict = SECOND();
    DEOPT_IF(!PyDict_CheckExact(dict), BINARY_SUBSCR);
    STAT_INC(BINARY_SUBSCR, hit);
    PyObject *sub = TOP(); // Leave sub on stack for binary_subscr_error
    PyObject *res = PyDict_GetItemWithError(dict, sub);
    if (res == NULL) {
        goto binary_subscr_error;
    }
    Py_INCREF(res);
    STACK_SHRINK(1);
    Py_DECREF(sub);
    SET_TOP(res);
    Py_DECREF(dict);
    DISPATCH();
}
```